pkg/master/controller.go

```diff
@@ -296,13 +303,104 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
+				collision = true
+			}
+		}
+		if !collision {
+			newEndpointPorts = append(newEndpointPorts, endpointPort)
+		}
+	}
+	return append(newEndpointPorts, missingPorts...)
+}
+
+// * updateMasterServiceIfNeeded sets service attributes for the
+//     given apiserver service.
+// * updateMasterServiceIfNeeded expects that the service object it
+//     manages will be managed only by updateMasterServiceIfNeeded;
+//     therefore, to understand this, you need only understand the
+//     requirements and the body of this function.
+// * updateMasterServiceIfNeeded ensures that the correct ports are
+//     present on the service but will not stomp existing ports.
+//
+// Requirements:
+// * All apiservers MUST use updateMasterServiceIfNeeded and only
+//     updateMasterServiceIfNeeded to manage service attributes
+// * updateMasterService is called periodically from all apiservers.
+func updateMasterServiceIfNeeded(svc *api.Service, servicePorts []api.ServicePort, serviceType api.ServiceType) (s *api.Service, updated bool) {
+	// Determine if the service is in the format we expect
+	// (servicePorts are present and service type matches)
+	formatCorrect, missingPorts := checkServiceFormat(svc, servicePorts, serviceType)
+	if formatCorrect {
+		return svc, false
+	}
+	if len(missingPorts) > 0 {
+		// Update service ports in context, replacing
+		// conflicting ports and appending missing ports.
+		svc.Spec.Ports = filterServicePorts(svc.Spec.Ports, missingPorts)
+	}
+	svc.Spec.Type = serviceType
+	return svc, true
+}
+
+// Determine if the service is in the correct format updateMasterServiceIfNeeded
+// expects (servicePorts are present and service type matches). Ports which are
+// missing are returned so that they can added.
+func checkServiceFormat(s *api.Service, ports []api.ServicePort, serviceType api.ServiceType) (formatCorrect bool, missingPorts []api.ServicePort) {
+	formatCorrect = true
+	for _, port := range ports {
+		contains := false
+		for _, servicePort := range s.Spec.Ports {
+			if port == servicePort {
+				contains = true
+			}
+		}
+		if !contains {
+			formatCorrect = false
+			missingPorts = append(missingPorts, port)
+		}
+	}
+	if s.Spec.Type != serviceType {
+		formatCorrect = false
+	}
+	return formatCorrect, missingPorts
+}
+
+// filterServicePorts combines missing service ports with existing service ports.
+// Missing ports take precedence over existing ports when a collision occurs.
+func filterServicePorts(servicePorts []api.ServicePort, missingPorts []api.ServicePort) []api.ServicePort {
+	newServicePorts := servicePorts[:0]
+	for _, servicePort := range servicePorts {
+		collision := false
+		for _, missingPort := range missingPorts {
+			if (missingPort.Name == servicePort.Name) || (missingPort.Port == servicePort.Port && missingPort.Protocol == servicePort.Protocol) || (missingPort.TargetPort == servicePort.TargetPort && missingPort.Protocol == servicePort.Protocol) {
```

Comment:
looks like name is optional if there's a single port? that means if this starts with a single nameless non-conflicting port, and appends other named non-conflicting ports, the service will fail validation because the original port had no name.